hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/qjournal/server/Journal.java

```diff
@@ -750,10 +750,17 @@ public GetJournaledEditsResponseProto getJournaledEdits(long sinceTxId,
           "is a requirement to fetch journaled edits via RPC. Please enable " +
           "it via " + DFSConfigKeys.DFS_HA_TAILEDITS_INPROGRESS_KEY);
     }
-    if (sinceTxId > getHighestWrittenTxId()) {
+    long highestTxId = getHighestWrittenTxId();
+    if (sinceTxId == highestTxId + 1) {
       // Requested edits that don't exist yet; short-circuit the cache here
       metrics.rpcEmptyResponses.incr();
       return GetJournaledEditsResponseProto.newBuilder().setTxnCount(0).build();
+    } else if (sinceTxId > highestTxId + 1) {
```

Comment:
I'm still a little wary about special-casing this. What's the concern with throwing the exception in in the `highestTxId + 1` case -- is it about the perf overhead of creating the exception? If so we could use a cached exception:

```java
private static final NewerTxnIdException NEW_TXN_ID_EXCEPTION = new JournaledEditsCache.NewerTxnIdException("...");

...

public GetJournaledEditsResponseProto getJournaledEdits(...) {
  ...
  if (sinceTxId > getHighestWrittenTxId()) {
    throw NEW_TXN_ID_EXCEPTION;
  }
  ...
}
```

The expensive part of throwing exceptions is populating the stack trace, which can be avoided like this.